lsp.go

```diff
@@ -144,6 +148,144 @@ func (s *Server) handleInitialize(
 	if err != nil {
 		log.Err(err).Msg("failed to notify client")
 	}
+	return nil
+}
+
+func (s *Server) handleDocumentOpen(
+	ctx context.Context,
+	reply jsonrpc2.Replier,
+	r jsonrpc2.Request,
+	log zerolog.Logger,
+) error {
+	var params lsp.DidOpenTextDocumentParams
+	if err := json.Unmarshal(r.Params(), &params); err != nil {
+		log.Error().Err(err).Msg("failed to unmarshal params")
+		return err
+	}
+
+	fname := params.TextDocument.URI.Filename()
+	content := params.TextDocument.Text
+
+	err := checkFile(fname, content)
+	return s.sendDiagnostics(ctx, params.TextDocument.URI, err)
+}
+
+func (s *Server) handleDocumentChange(
+	ctx context.Context,
+	reply jsonrpc2.Replier,
+	r jsonrpc2.Request,
+	log zerolog.Logger,
+) error {
+	var params lsp.DidChangeTextDocumentParams
+	if err := json.Unmarshal(r.Params(), &params); err != nil {
+		log.Error().Err(err).Msg("failed to unmarshal params")
+		return err
+	}
+
+	if len(params.ContentChanges) != 1 {
+		err := fmt.Errorf("unexpected content change length: %d", len(params.ContentChanges))
+		log.Error().Err(err).Send()
+		return err
+	}
+
+	content := params.ContentChanges[0].Text
+	fname := params.TextDocument.URI.Filename()
+
+	err := checkFile(fname, content)
+	return s.sendDiagnostics(ctx, params.TextDocument.URI, err)
+}
+
+func (s *Server) sendDiagnostics(ctx context.Context, uri lsp.URI, err error) error {
+	diags := []lsp.Diagnostic{}
+
+	if err != nil {
+		e, ok := err.(*errors.Error)
+		if ok {
+			log.Debug().Str("error", e.Detailed()).Msg("failed to parse hcl directory")
+
+			fileRange := lsp.Range{}
+			fileRange.Start.Line = uint32(e.FileRange.Start.Line) - 1
+			fileRange.Start.Character = uint32(e.FileRange.Start.Column) - 1
+			fileRange.End.Line = uint32(e.FileRange.End.Line) - 1
+			fileRange.End.Character = uint32(e.FileRange.End.Column) - 1
```

Comment:
hmm LSP starts on line/collumn 0 and then HCL/Terramate starts at 1:1 ? :thinking: